<html><head><meta charset='utf-8'><meta name='viewport' content='width=device-width, initial-scale=1'>
<meta name='applicable-device' content='pc'><meta name='keywords' content='电脑,电脑讲解,电脑技术,编程,电脑故障维修C++中函数异常规格的说明' />
<script src='../../highlight/highlight.pack.js'></script>
<link rel='stylesheet' type='text/css' href='../../highlight/styles/monokai.css'/>

<link rel='stylesheet' href='../../fenxiang/dist/css/share.min.css'>
<script src='../../fenxiang/src/js/social-share.js'></script>
<script src='../../fenxiang/src/js/qrcode.js'></script>

</head><body><script>hljs.initHighlightingOnLoad();</script><script>
var system ={};  
var p = navigator.platform;       
system.win = p.indexOf('Win') == 0;  
system.mac = p.indexOf('Mac') == 0;  
system.x11 = (p == 'X11') || (p.indexOf('Linux') == 0);     
if(system.win||system.mac||system.xll){
document.write("<link href='../css/3.css' rel='stylesheet' type='text/css'>");}else{ document.write("<link href='../css/3wap.css' rel='stylesheet' type='text/css'>");}</script><script src='../../js/3.js'></script><div class='div2'><div class='heading_nav'><ul><div><li><a href='../../index.html'>首页</a></li>
</div><div onclick='hidden1()' >分享</div>
</ul></div></div>
<div id='heading_nav2'> 
<li class='row' >
<div class='social-share' data-mode='prepend'><a href='javascript:' class='social-share-icon icon-heart'></a></div></li></div><script charset='utf-8' src='../../3/js/hengfu.js'></script><script charset='utf-8' src='../../3/js/hengfu2.js'></script><hr><div class='div1'><div class='biaoti'><center>C++中函数异常规格的说明</center></div><div class='banquan'>原文出处:本文由博客园博主子宇24提供。<br/>
原文连接:https://www.cnblogs.com/dishengAndziyu/p/10920150.html</div><br>
    <p align="left">1，本文介绍一个新的概念，它是一个重要的概念，并且是 C++ 中的一个高级主题；</p>
<p align="left">&nbsp;</p>
<p align="left">2，问题：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，如何判断一个函数（不是自己写的，有可能是第三方库中的函数）是否会抛出异常，以及抛出那些异常？</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，学习了异常处理，我们在调用一个函数的时候，就应该知道这个函数会不会抛出异常，如果会，抛那些：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，如果是第三方库的函数，只有函数声明，没有函数实现，则也不知道会不会抛出异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3，查看函数文档也是不错的方法，但是如果手头文档和真实的库的版本不匹配，有可能查到的东西就是不准确的；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">3，异常规格说明：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，C++ 提供语法用于声明函数所抛出的异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，异常声明作为函数声明的修饰符，写在参数列表后面：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，/* 可能抛出任何异常 */</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void func1();</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; /* 只能抛出的异常类型：char 和 int */</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void func2() throw(char, int);</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; /* 不抛出任何异常 */</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void func3() throw();</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p>
<p align="left">4，异常规格说明的意义：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，提示函数调用者必须做好异常处理的准备；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，如果想知道调用的函数会抛出哪些类型的异常时，只用打开头文件看看这个函数是怎么声明的就可以了；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，提示函数的维护者不要抛出其它异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，异常规格说明是函数接口的一部分；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，用于说明这个函数如何正确的使用；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">5，如果抛出的异常不在声明列表中，会发生什么？</p>
<p align="left">&nbsp;</p>
<p align="left">6，下面的代码输出什么？</p>
<p align="center"><img src="./images/C++中函数异常规格的说明0.png" alt="" /></p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">7，异常规格之外的异常编程实验：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;"> 1</span> #include &lt;iostream&gt;
<span style="color: #008080;"> 2</span> 
<span style="color: #008080;"> 3</span> <span style="color: #0000ff;">using</span> <span style="color: #0000ff;">namespace</span><span style="color: #000000;"> std;
</span><span style="color: #008080;"> 4</span> 
<span style="color: #008080;"> 5</span> <span style="color: #0000ff;">void</span> func() <span style="color: #0000ff;">throw</span>(<span style="color: #0000ff;">int</span><span style="color: #000000;">)
</span><span style="color: #008080;"> 6</span> <span style="color: #000000;">{
</span><span style="color: #008080;"> 7</span>     cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">func()</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;"> 8</span>     cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;"> 9</span>     
<span style="color: #008080;">10</span>     <span style="color: #0000ff;">throw</span> <span style="color: #800000;">'</span><span style="color: #800000;">c</span><span style="color: #800000;">'</span><span style="color: #000000;">;
</span><span style="color: #008080;">11</span> <span style="color: #000000;">}
</span><span style="color: #008080;">12</span> 
<span style="color: #008080;">13</span> <span style="color: #0000ff;">int</span><span style="color: #000000;"> main()
</span><span style="color: #008080;">14</span> <span style="color: #000000;">{
</span><span style="color: #008080;">15</span>     <span style="color: #0000ff;">try</span> 
<span style="color: #008080;">16</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">17</span> <span style="color: #000000;">        func();
</span><span style="color: #008080;">18</span> <span style="color: #000000;">    } 
</span><span style="color: #008080;">19</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">int</span><span style="color: #000000;">) 
</span><span style="color: #008080;">20</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">21</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(int)</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;">22</span>         cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">23</span> <span style="color: #000000;">    } 
</span><span style="color: #008080;">24</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">char</span><span style="color: #000000;">) 
</span><span style="color: #008080;">25</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">26</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(char)</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;">27</span>         cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">28</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">29</span> 
<span style="color: #008080;">30</span>     <span style="color: #0000ff;">return</span> <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;">31</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp; 1，BCC 显示：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; Abnormal program termination</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，g++ 显示：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，terminate called after throwing an instance of 'char'</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，已放弃</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，VC 2010 显示：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，catch(char)</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">8，unexpected() 函数说明：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，函数抛出的异常不在规格说明中，全局 unexpected() 被调用；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，默认的 unexpected() 函数会调用全局的 terminate() 函数；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，这是 BCC 和 g++ 编译器的行为；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，可以自定义函数替换默认的 unexpected() 函数实现；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，VC 2010 编译器的也不遵循行为；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 4，注意：不是所有的 C++ 编译器都支持这个标准行为；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，在异常处理这个技术点上面，编译器实现是有差异的，如果我们未来的项目中确实的要用到函数的异常规格说明时，我们最好的写一个小的测试程序，看看当前项目里面所使用的 C++ 编译器在这一个技术点的行为是怎样的，有没有遵循标准的 C++ 规范；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">9，unexpected() 函数的替换：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，自定义一个无返回值无参数的函数：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，能够再次抛出异常：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，当异常符合触发函数的异常规格说明时，恢复程序执行；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，见 本文10 中程序 throw 1；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，否则，调用全局 terminate() 函数结束程序；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，调用 set_unexpected() 设置自定义的异常函数：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，参数类型为 void(*)()；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，返回值为默认的 unexpected() 函数入口地址；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p>
<p align="left">10，自定义 unexpected() 函数编程实验：</p>
<div class="cnblogs_code">
<pre><code><span style="color: #008080;"> 1</span> #include &lt;iostream&gt;
<span style="color: #008080;"> 2</span> #include &lt;cstdlib&gt;
<span style="color: #008080;"> 3</span> #include &lt;exception&gt;
<span style="color: #008080;"> 4</span> 
<span style="color: #008080;"> 5</span> <span style="color: #0000ff;">using</span> <span style="color: #0000ff;">namespace</span><span style="color: #000000;"> std;
</span><span style="color: #008080;"> 6</span> 
<span style="color: #008080;"> 7</span> <span style="color: #0000ff;">void</span><span style="color: #000000;"> my_unexpected()
</span><span style="color: #008080;"> 8</span> <span style="color: #000000;">{
</span><span style="color: #008080;"> 9</span>     cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">void my_unexpected()</span><span style="color: #800000;">"</span> &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">10</span>     <span style="color: #008000;">//</span><span style="color: #008000;"> exit(1);  </span><span style="color: #008000;">//</span><span style="color: #008000;"> 退出；</span>
<span style="color: #008080;">11</span>     <span style="color: #0000ff;">throw</span> <span style="color: #800080;">1</span><span style="color: #000000;">;
</span><span style="color: #008080;">12</span> <span style="color: #000000;">}
</span><span style="color: #008080;">13</span> 
<span style="color: #008080;">14</span> <span style="color: #0000ff;">void</span> func() <span style="color: #0000ff;">throw</span>(<span style="color: #0000ff;">int</span><span style="color: #000000;">)
</span><span style="color: #008080;">15</span> <span style="color: #000000;">{
</span><span style="color: #008080;">16</span>     cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">func()</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;">17</span>     cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">18</span>     
<span style="color: #008080;">19</span>     <span style="color: #0000ff;">throw</span> <span style="color: #800000;">'</span><span style="color: #800000;">c</span><span style="color: #800000;">'</span><span style="color: #000000;">;
</span><span style="color: #008080;">20</span> <span style="color: #000000;">}
</span><span style="color: #008080;">21</span> 
<span style="color: #008080;">22</span> <span style="color: #0000ff;">int</span><span style="color: #000000;"> main()
</span><span style="color: #008080;">23</span> <span style="color: #000000;">{
</span><span style="color: #008080;">24</span> <span style="color: #000000;">    set_unexpected(my_unexpected);
</span><span style="color: #008080;">25</span>     
<span style="color: #008080;">26</span>     <span style="color: #0000ff;">try</span> 
<span style="color: #008080;">27</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">28</span> <span style="color: #000000;">        func();
</span><span style="color: #008080;">29</span> <span style="color: #000000;">    } 
</span><span style="color: #008080;">30</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">int</span><span style="color: #000000;">) 
</span><span style="color: #008080;">31</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">32</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(int)</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;">33</span>         cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">34</span> <span style="color: #000000;">    } 
</span><span style="color: #008080;">35</span>     <span style="color: #0000ff;">catch</span>(<span style="color: #0000ff;">char</span><span style="color: #000000;">) 
</span><span style="color: #008080;">36</span> <span style="color: #000000;">    {
</span><span style="color: #008080;">37</span>         cout &lt;&lt; <span style="color: #800000;">"</span><span style="color: #800000;">catch(char)</span><span style="color: #800000;">"</span><span style="color: #000000;">;
</span><span style="color: #008080;">38</span>         cout &lt;&lt;<span style="color: #000000;"> endl;
</span><span style="color: #008080;">39</span> <span style="color: #000000;">    }
</span><span style="color: #008080;">40</span> 
<span style="color: #008080;">41</span>     <span style="color: #0000ff;">return</span> <span style="color: #800080;">0</span><span style="color: #000000;">;
</span><span style="color: #008080;">42</span> }</pre>
</div>
<p align="left">&nbsp;&nbsp;&nbsp; 1，exit(1) 结果：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，g++ 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void my_unexpected()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，BCC 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void my_unexpected()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3，VC 2010 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; catch(char)&nbsp; // 未有遵循标准 C++ 规范；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，throw 1 结果：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，g++ 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void my_unexpected()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; catch(int)&nbsp; //与触发函数 func() 的异常规格符合，于是程序恢复执行；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，BCC 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; void my_unexpected()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; catch(int)</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3，VC 2010 编译器：</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; func()</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp; catch(char)&nbsp; // 未有遵循标准 C++ 规范，直接捕获，然后处理；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，如果以后项目开发，会使用函数异常说明这个技术点，最好在项目开发前写一些小程序测试下当前使用的 C++ 编译器有没有很好的遵循 C++ 的规范；</p>
<p align="left">&nbsp;&nbsp;&nbsp;</p>
<p align="left">11，小结：</p>
<p align="left">&nbsp;&nbsp;&nbsp; 1，C++ 中的函数可以声明异常规则说明；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 2，异常规格说明可以看作接口的一部分；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 3，函数抛出的异常不在规格说明中，unexpected() 被调用；</p>
<p align="left">&nbsp;&nbsp;&nbsp; 4，unexpected() 中能够再次抛出异常；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 1，异常能够匹配，恢复程序的执行；</p>
<p align="left">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 2，否则，调用 terminate() 结束程序；</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; 3，un_expected() 函数是正确处理异常的最后机会，如果没有抓住，terminate() 函数会被调用，当前程序以异常告终；</p>
</div>
</div><hr><script charset='utf-8' src='../../js/sming.js'></script></body></html>